esp-hal/src/i2s/master/camera.rs

```diff
@@ -0,0 +1,379 @@
+        I2S::din2_signal().connect_to(d2);
+        I2S::din3_signal().connect_to(d3);
+        I2S::din4_signal().connect_to(d4);
+        I2S::din5_signal().connect_to(d5);
+        I2S::din6_signal().connect_to(d6);
+        I2S::din7_signal().connect_to(d7);
+
+        self
+    }
+
+    /// FIXME: docs
+    pub fn with_ws<PIN: InputPin>(self, pin: impl Peripheral<P = PIN>) -> Self {
+        crate::into_mapped_ref!(pin);
+        pin.init_input(Pull::None);
+        I2S::ws_in_signal().connect_to(pin);
+        self
+    }
+
+    /// FIXME: docs
+    pub fn with_vsync<PIN: InputPin>(self, pin: impl Peripheral<P = PIN>) -> Self {
+        crate::into_mapped_ref!(pin);
+        pin.init_input(Pull::None);
+        I2S::v_sync_signal().connect_to(pin);
+        self
+    }
+
+    /// FIXME: docs
+    pub fn with_hsync<PIN: InputPin>(self, pin: impl Peripheral<P = PIN>) -> Self {
+        crate::into_mapped_ref!(pin);
+        pin.init_input(Pull::None);
+        I2S::h_sync_signal().connect_to(pin);
+        self
+    }
+
+    /// FIXME: docs
+    pub fn with_henable<PIN: InputPin>(self, pin: impl Peripheral<P = PIN>) -> Self {
+        crate::into_mapped_ref!(pin);
+        pin.init_input(Pull::None);
+        I2S::h_enable_signal().connect_to(pin);
+        self
+    }
+}
+
+impl<'d, I2S> Camera<'d, I2S>
+where
+    I2S: RegisterAccess,
+{
+    fn start_rx_transfer<'t, RXBUF>(
+        &'t mut self,
+        buf: &'t mut RXBUF,
+        len: usize,
+    ) -> Result<(), Error>
+    where
+        RXBUF: DmaRxBuffer,
+    {
+        if len % 4 != 0 {
+            return Err(Error::IllegalArgument);
+        }
+
+        // Reset RX unit and RX FIFO
+        self._i2s.reset_rx();
+
+        // Enable corresponding interrupts if needed
+
+        // configure DMA outlink
+        unsafe {
+            self.rx_channel
+                .prepare_transfer(self._i2s.dma_peripheral(), buf)
+                .and_then(|_| self.rx_channel.start_transfer())?;
+        }
+
+        // set I2S_RX_STOP_EN if needed
+
+        // start: set I2S_RX_START
+        self._i2s.rx_start(len);
+        Ok(())
+    }
+
+    /// FIXME: docs
+    pub fn receive<BUF: DmaRxBuffer>(
+        mut self,
+        mut buf: BUF,
+        len: usize,
+    ) -> Result<CameraTransfer<'d, I2S, BUF>, Error> {
+        self.start_rx_transfer(&mut buf, len)?;
+        Ok(CameraTransfer {
+            camera: ManuallyDrop::new(self),
+            buffer_view: ManuallyDrop::new(buf.into_view()),
+        })
+    }
+}
+
+/// Represents an ongoing (or potentially stopped) transfer from the Camera to a
+/// DMA buffer.
+pub struct CameraTransfer<'d, I2S: DmaEligible + RegisterAccess, BUF: DmaRxBuffer> {
+    camera: ManuallyDrop<Camera<'d, I2S>>,
+    buffer_view: ManuallyDrop<BUF::View>,
+}
+
+impl<'d, I2S: RegisterAccess, BUF: DmaRxBuffer> CameraTransfer<'d, I2S, BUF> {
+    /// Returns true when [Self::wait] will not block.
+    pub fn is_done(&self) -> bool {
```

Comment:
The booleans of this function are inverted, which is why the DMA wasn't working for me.

After fixing this function I do get a valid JPEG